### PR TITLE
scipified the COM function _get_molecules

### PR DESCRIPTION
**Pull request by Harry-Rich**
Added function for _get_molecules and implemented in pymatgen parser. 

Still need to implement within other parser and implement get_framework for single molecules

### CODE REVIEW

File: kinisi/parser.py [hunk 1]
```diff
@@ -147,19 +147,33 @@ class PymatgenParser(Parser):
...
+        self,
+        structures: List['pymatgen.core.structure.Structure'],
+        specie: Union['pymatgen.core.periodic_table.Element', 'pymatgen.core.periodic_table.Specie'],
+        time_step: sc.Variable,
+        step_skip: sc.Variable,
+        dt: sc.Variable = None,
+        dimension: str = 'xyz',
+        distance_unit: sc.Unit = sc.units.angstrom,
+        progress: bool = True,
+        specie_indices: sc.Variable = None,
```
Comment (arm61) [inline kinisi/parser.py:160]: I think that `progress` should be the final keyword argument. Also, can we add something to the docstring for the new arguments?

File: kinisi/parser.py [hunk 2]
```diff
@@ -147,19 +147,33 @@ class PymatgenParser(Parser):
...
         structure, coords, latt = self.get_structure_coords_latt(structures, progress)
-        indices, drift_indices = self.get_indices(structure, specie)
+
+        if specie is not None:
+            indices, drift_indices = self.get_indices(structure, specie)
+        elif isinstance(specie_indices, sc.Variable):
+            if len(specie_indices.dims) > 1:
+                coords, indices, drift_indices = _get_molecules(structure, coords, specie_indices, masses)
+            else:
+                raise TypeError('to implement')
```
Comment (arm61) [inline kinisi/parser.py:173]: Can you add this to the TODO-scipp.md file?

File: kinisi/parser.py [hunk 3]
```diff
@@ -347,3 +361,77 @@ def get_indices(
...
+    """
+
+    s_coords = coords[:, indices]
+    theta = s_coords * (2 * np.pi)
+    xi = np.cos(theta)
+    zeta = np.sin(theta)
+    xi_bar = np.average(xi, axis=-2, weights=weights)
+    zeta_bar = np.average(zeta, axis=-2, weights=weights)
+    theta_bar = np.arctan2(zeta_bar, xi_bar)
+    new_s_coords = theta_bar / (2 * np.pi)
```
Comment (arm61) [inline kinisi/parser.py:435]: If we are keeping this non-scipp for now, let's put something in the TODO-scipp.md file.

**Review by arm61 — APPROVED** [review]

File: kinisi/parser.py [hunk 4]
```diff
@@ -347,3 +361,77 @@ def get_indices(
...
+    and Tuple containing: indices for centers used in the calculation 
+    of the diffusion and indices of framework atoms.
+    """
+    drift_indices = []
+    try:
+        indices.values = indices.values - 1
+    except:
+        raise ValueError('Molecules must be of same length')
+
+    n_molecules = len(indices.dims)
```
Comment (arm61) [inline kinisi/parser.py:390]: I think that this is a bug and should read:
```suggestion
    n_molecules = indices.shape[0]
```
To reproduce the current implementation.

File: kinisi/parser.py [hunk 5]
```diff
@@ -274,13 +274,23 @@ def __init__(self,
...
-        indices, drift_indices = self.get_indices(structure, specie)
+        if specie is not None:
+            indices, drift_indices = self.get_indices(structure, specie)
+        elif isinstance(specie_indices, sc.Variable):
+            if len(specie_indices.dims) > 1:
+                coords, indices, drift_indices = _get_molecules(structure, coords, specie_indices, masses)
+            else:
+                indices, drift_indices = _get_framework(structure, specie_indices)
+        else:
+            raise TypeError('Unrecognized type for specie or specie_indices, specie_indices must be a sc.array')
```
Comment (arm61) [inline kinisi/parser.py:293]: We should maybe have some error-checking for the dimension names?

File: kinisi/parser.py [hunk 6]
```diff
@@ -347,3 +371,106 @@ def get_indices(
...
+    # This allows the dimensions of the indices to be any word, paired with 'atom'.
+    dims_id = [i for i in indices.dims if i != 'atom'][0]
+    xi_bar = (weights * xi).sum(dim=dims_id) / weights.sum(dim=dims_id)
+    zeta_bar = (weights * zeta).sum(dim=dims_id) / weights.sum(dim=dims_id)
+    theta_bar = sc.atan2(y=-zeta_bar, x=-xi_bar) + np.pi * sc.units.rad
+    new_s_coords = theta_bar / (2 * np.pi * (sc.units.rad / sc.units.angstrom))
+    return new_s_coords
+
+
+def _get_framework(structure: "ase.atoms.Atoms" or "pymatgen.core.structure.Structure"
```
Comment (arm61) [inline kinisi/parser.py:454]: Can we use the `Union` thing instead of `or`?

File: kinisi/parser.py [hunk 7]
```diff
@@ -347,3 +371,106 @@ def get_indices(
...
+    s_coords = sc.fold(coords['atom', indices.values.flatten()], 'atom', dims=indices.dims, shape=indices.shape)
+    theta = s_coords * (2 * np.pi * (sc.units.rad / sc.units.angstrom))
+    xi = sc.cos(theta)
+    zeta = sc.sin(theta)
+    # This allows the dimensions of the indices to be any word, paired with 'atom'.
+    dims_id = [i for i in indices.dims if i != 'atom'][0]
+    xi_bar = (weights * xi).sum(dim=dims_id) / weights.sum(dim=dims_id)
+    zeta_bar = (weights * zeta).sum(dim=dims_id) / weights.sum(dim=dims_id)
+    theta_bar = sc.atan2(y=-zeta_bar, x=-xi_bar) + np.pi * sc.units.rad
+    new_s_coords = theta_bar / (2 * np.pi * (sc.units.rad / sc.units.angstrom))
```
Comment (arm61) [inline kinisi/parser.py:450]: Should this be the `distance_unit` too then?

**Comment by arm61** [comment]
Do we want to start thinking about tests...?

File: kinisi/parser.py (hunk 7 above)
Comment (Harry-Rich) [reply kinisi/parser.py:450]: yes

Comment (arm61) [reply kinisi/parser.py:450]: Can you have a look to see if there is anywhere else that needs changed too?

File: kinisi/parser.py [hunk 8]
```diff
@@ -427,7 +428,7 @@ def _get_molecules(structure: Union["ase.atoms.Atoms", "pymatgen.core.structure.
 
 
 def _calculate_centers_of_mass(coords: VariableLikeType, weights: VariableLikeType,
-                               indices: VariableLikeType) -> VariableLikeType:
+                               indices: VariableLikeType, distance_unit:VariableLikeType) -> VariableLikeType:
```
Comment (arm61) [inline kinisi/parser.py:431]: Can we double check that this is the correct type of for the distance unit?

**Review by arm61 — APPROVED** [review]

File: kinisi/parser.py (hunk 8 above)
Comment (Harry-Rich) [reply kinisi/parser.py]: Sorry should be sc.Unit right

**Comment by arm61** [comment]
Love a Friday afternoon/evening merge!

**Pull request merged**